crates/sui-core/src/consensus_adapter.rs

```diff
@@ -184,6 +186,78 @@ impl SubmitToConsensus for TransactionsClient<sui_network::tonic::transport::Cha
     }
 }
 
+/// A Narwhal client that instantiates LocalNarwhalClient lazily.
+pub struct LazyNarwhalClient {
+    /// Outer ArcSwapOption allows initialization after the first the connection to Narwhal.
+    /// Inner ArcSwap allows Narwhal restarts across epoch changes.
+    client: ArcSwapOption<ArcSwap<LocalNarwhalClient>>,
+    addr: Multiaddr,
+}
+
+impl LazyNarwhalClient {
+    /// Lazily instantiates LocalNarwhalClient keyed by the address of the Narwhal worker.
+    pub fn new(addr: Multiaddr) -> Self {
```

Comment:
Do we accept the `addr` to make the implementation agnostic? Otherwise I would say using the `worker_id` would be more straightforward since we have this info available

Reply:
Worker ID is less accessible in consensus adapter and Sui tests in general. But this will be refactored as well.